docs/source/task_guides/lora_based_methods.md

```diff
@@ -0,0 +1,345 @@
+
+```py
+from peft import LoHaConfig, get_peft_model
+
+config = LoHaConfig(
+    r=16,
+    alpha=16,
+    target_modules=["query", "value"],
+    module_dropout=0.1,
+    modules_to_save=["classifier"],
+)
+model = get_peft_model(model, config)
+model.print_trainable_parameters()
+"trainable params: 1,257,317 || all params: 87,133,642 || trainable%: 1.4429753779831676"
+```
+
+</hfoption>
+<hfoption id="LoKr">
+
+[LoKr](../conceptual_guides/adapter#low-rank-kronecker-product-lokr) expresses the weight update matrix as a decomposition of a Kronecker product, creating a block matrix that is able to preserve the rank of the original weight matrix. The size of the smaller matrices are determined by its *rank* or `r`. You'll also want to specify the `target_modules` which determines where the smaller matrices are inserted. For this guide, you'll target the *query* and *value* matrices of the attention blocks. Other important parameters to set are `alpha` (scaling factor), and `modules_to_save` (the modules apart from the LoKr layers to be trained and saved). All of these parameters - and more - are found in the [`LoKrConfig`].
+
+```py
+from peft import LoKrConfig, get_peft_model
+
+config = LoKrConfig(
+    r=16,
+    alpha=16,
+    target_modules=["query", "value"],
+    module_dropout=0.1,
+    modules_to_save=["classifier"],
+)
+model = get_peft_model(model, config)
+model.print_trainable_parameters()
+"trainable params: 116,069 || all params: 87,172,042 || trainable%: 0.13314934162033282"
+```
+
+</hfoption>
+<hfoption id="AdaLoRA">
+
+[AdaLoRA](../conceptual_guides/adapter#adaptive-low-rank-adaptation-adalora) efficiently manages the LoRA parameter budget by assigning important weight matrices more parameters and pruning less important ones. In contrast, LoRA evenly distributes parameters across all modules. You can control the average desired *rank* or `r` of the matrices, and which modules to apply AdaLoRA to with `target_modules`. Other important parameters to set are `lora_alpha` (scaling factor), and `modules_to_save` (the modules apart from the AdaLoRA layers to be trained and saved). All of these parameters - and more - are found in the [`AdaLoraConfig`].
```

Comment:
There is a bit of an issue with the AdaLora example. For AdaLora, we need to call the method `update_and_allocate` during each training step, otherwise, the adaptation step is not performed. This requires to either write a custom training loop or to subclass the `Trainer` to add this step. An example of a custom training loop that calls this method is here:

https://github.com/huggingface/peft/blob/912ad41e96e03652cabf47522cd876076f7a0c4f/examples/conditional_generation/peft_adalora_seq2seq.py#L120

Given this guide, as is, the reader may not be aware that this is required and thus use AdaLora incorrectly. These solutions come to mind:

1. Don't provide the AdaLora example. Instead, refer to the existing examples with custom training loops.
2. Don't use `Trainer`, instead use a custom training loop that calls this method when AdaLora is being used.
3. Subclass `Trainer` to perform this step if AdaLora is being used. I don't know enough about `Trainer` to be really sure how to do that correctly.

Reply:
Great point! I mention `update_and_allocate` + link to a custom training loop in a `Warning` in the training section for more visibility. I think we should still keep the example of showing the `AdaLoraConfig` though so users still know how to set it up.